pkg/snmp/traps/server.go

```diff
@@ -0,0 +1,174 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2020 Datadog, Inc.
+
+package traps
+
+import (
+	"net"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/soniah/gosnmp"
+)
+
+/*
+* Traps pipeline diagram
+* ----------------------
+*
+* (devices) --[udp:port1]-!-> (Listener 1) \
+*                         !                 (TrapServer) --[chan]--> (Forwarder) --[chan]--> (Rest of the logs pipeline...)
+* (devices) --[udp:portN]-!-> (Listener N) /
+*
+* ! = delimitation between the outside world and the Agent
+ */
+
+// SnmpPacket is the type of packets yielded by server listeners.
+type SnmpPacket struct {
+	Content *gosnmp.SnmpPacket
+	Addr    *net.UDPAddr
+}
+
+// OutputChannel is the type of the server output channel.
+type OutputChannel = chan *SnmpPacket
+
+// TrapServer runs multiple SNMP traps listeners.
+type TrapServer struct {
+	Started bool
+
+	listeners   []TrapListener
+	running     []bool
+	output      OutputChannel
+	stopTimeout time.Duration
+}
+
+var (
+	outputChannelSize = 100
+	// RunningServer holds a reference to the trap server instance running in the Agent.
+	RunningServer *TrapServer
+)
+
+// NewTrapServer configures and returns a running SNMP traps server.
+// Misconfigured listeners are be dropped, allowing valid listeners to start in all cases.
+func NewTrapServer() (*TrapServer, error) {
+	var configs []TrapListenerConfig
+	err := config.Datadog.UnmarshalKey("snmp_traps_listeners", &configs)
+	if err != nil {
+		return nil, err
+	}
+
+	defaultBindHost := config.Datadog.GetString("bind_host")
+	stopTimeout := config.Datadog.GetDuration("snmp_traps_stop_timeout") * time.Second
+
+	output := make(OutputChannel, outputChannelSize)
+	listeners := make([]TrapListener, 0, len(configs))
+	running := make([]bool, 0, len(configs))
```

Comment:
Any reason `running` is not a boolean field on `TrapListener` ?

Reply:
This is initially what I did: `TrapListener` would manage a `.Running` flag that it sets and unsets in `.Listen()`.

But I ran into race conditions in tests. Because `.Listen()` is run in a goroutine, if the rest of operations are sync (like in a test where I just inspect the number of running listeners) then the flag is only set after those operations are done.

So even if `NewTrapServer()` returned a started server, the goroutines hadn't started and the flags weren't set. So instead since the server is the one managing the lifecycle of listeners, I put the "keep track of registered/running/failed listeners" logic there.

Reply:
FYI, pushed the commit that flips how this state is tracked (running listeners -> failed listeners). Let me know if how this looks.